src/sentry/sentry_metrics/querying/api.py

```diff
@@ -289,44 +348,130 @@ def _build_request(self, query: MetricsQuery) -> Request:
-        # By default we want to execute totals.
-        self._scheduled_queries.append((query, with_totals))
+    def schedule(self, query: MetricsQuery, group_bys: Optional[Sequence[str]]):
+        executable_query = ExecutableQuery(
+            query=query,
+            group_bys=group_bys,
+            # For now, we execute both queries independently of the query.
+            with_series=True,
+            with_totals=True,
+        )
+        self._scheduled_queries.append(executable_query)
+
+    def _derive_optimal_interval(self, result: ExecutionResult) -> int:
+        # First we determine as best effort how many groups were returned, just to estimate the cardinality. Keep in
+        # mind that this is an estimation, since it could be that the user has more than 10k groups which is highly
+        # unlikely but can happen. If that happens though, we might need to rethink snuba limits.
+        groups_number = len(result.groups)
+
+        # Second we compute the ideal number of intervals that can fit with a given number of groups. We round down
+        # since we want to work with integers.
+        intervals_number = math.floor(SNUBA_QUERY_LIMIT / groups_number)
+
+        # Third we compute the size of each interval in seconds, considering how many intervals we want in the time
+        # range of the query.
+        optimal_interval_size = math.floor(
+            (result.modified_end - result.modified_start).total_seconds() / intervals_number
+        )
+
+        # We want to get the biggest interval that is >= the optimal one. The idea for this heuristic is that we want
+        # to find the smallest interval possible that will give us fewer results than the maximum but since we don't
+        # want to use the mathematically optimal one, we try to find the closest given a set of choices. This is more
+        # of a product decision, since we want our customers to see the graphs with easily readable intervals.
+        for index, interval in enumerate(self._interval_choices):
+            if interval >= optimal_interval_size:
+                # We have to put the choice, otherwise we end up in an infinite recursion.
+                self._interval_choices.pop(index)
+                return interval
+
+        # If no suitable interval is found, we throw an error to unwind.
+        raise MetricsQueryExecutionError(
+            "Unable to find an interval to satisfy the query because too many results "
+            "are returned."
+        )
+
+    def _serial_execute(self) -> Sequence[ExecutionResult]:
+        if not self._scheduled_queries:
+            return []
+
+        first_query = self._scheduled_queries.pop(0)
+        first_result = self._execute(first_query)
+
+        # Case 1: we have fewer results that the limit. In this case we are free to run the follow-up queries under the
+        # assumption that data doesn't change much between queries.
```

Comment:
nit: We dont need the ordering of the cases
```suggestion
        # We have fewer results that the limit. In this case we are free to run the follow-up queries under the
        # assumption that data doesn't change much between queries.
```